hydroflow/src/scheduled/net/network_vertex.rs

```diff
@@ -18,6 +21,100 @@ pub type ReceiverId = u32;
+        &mut self,
+    ) -> (
+        u16,
+        OutputPort<VecHandoff<(ReceiverId, Message)>>,
+        OutputPort<VecHandoff<ReceiverId>>,
+        InputPort<VecHandoff<(u32, Message)>>,
+    ) {
+        let listener = TcpListener::bind("localhost:0").await.unwrap();
+        let port = listener.local_addr().unwrap().port();
+
+        let (mut conns, egress_edge) = self.add_egress_vertex();
+        // TODO(justin): figure out an appropriate buffer here.
+        let (incoming_send, incoming_messages) = futures::channel::mpsc::channel(1024);
+        let (mut incoming_conns, conns_recv) = futures::channel::mpsc::channel(1024);
+
+        // Listen to incoming connections and spawn a tokio task for each one,
+        // which feeds into the channel.
+        tokio::spawn(async move {
+            let mut conn_id = 0;
+            loop {
+                let (socket, _) = listener.accept().await.unwrap();
+                let (reader, writer) = socket.into_split();
+                conns.feed((conn_id, writer)).await.unwrap();
+                conns.flush().await.unwrap();
+                let mut reader = FramedRead::new(reader, LengthDelimitedCodec::new());
+                let mut incoming_send = incoming_send.clone();
+                let id = conn_id;
+                incoming_conns.feed(conn_id).await.unwrap();
+                incoming_conns.flush().await.unwrap();
+                tokio::spawn(async move {
+                    while let Some(msg) = reader.next().await {
+                        // TODO(justin): figure out error handling here.
+                        let msg = msg.unwrap();
+                        let msg = Message::decode(&msg.freeze());
+                        incoming_send.feed((id, msg)).await.unwrap();
+                    }
+                    // TODO(justin): The connection is closed, so we should
+                    // clean up its metadata somehow and issue a retraction for
+                    // the connection.
+                });
+                conn_id += 1;
+            }
+        });
+
+        let incoming_messages = self.add_input_from_stream(incoming_messages.map(Some));
+        let conns_recv = self.add_input_from_stream(conns_recv.map(Some));
+
+        (port, incoming_messages, conns_recv, egress_edge)
+    }
+
+    // TODO(justin): we should have a vertex that can dynamically connect to
+    // stuff based on data instead of this.
+    pub async fn connect_tcp<A>(
+        &mut self,
+        addr: A,
+    ) -> (
+        InputPort<VecHandoff<Message>>,
+        OutputPort<VecHandoff<Message>>,
+    )
+    where
+        A: ToSocketAddrs,
+    {
+        let stream = TcpStream::connect(addr).await.unwrap();
+
+        let (reader, writer) = stream.into_split();
+
+        let (mut incoming_send, incoming_messages) = futures::channel::mpsc::channel(1024);
+        let mut reader = FramedRead::new(reader, LengthDelimitedCodec::new());
+        tokio::spawn(async move {
```

Comment:
TcpStream/FramedRead etc implement Stream, would somehow chaining directly like this instead work?
```rust
let my_stream = tcp_stream.map(|msg| Message::decode(&msg.freeze()).ok());
self.add_input_from_stream(my_stream);
```

Reply:
nice, thats way better.